Custom root-path with custom non-application-root-path and access-log.enabled=true yields java.lang.NullPointerException

### Describe the bug

When using a custom `http.root-path` together with custom `http.non-application-root-path` and `http.access-log.enabled=true` starting a quarkus app fails withjava.lang.NullPointerException.

Example configuration which triggers the problem:
```
quarkus.http.root-path=/auth
quarkus.http.non-application-root-path=/q
quarkus.http.access-log.enabled=true
```

Note that this also happens with ealier quarkus versions. So if this is fixed it would be great to backport this :) (Quarkus 3.8.x maybe? :))

### Expected behavior

The app should not fail to start

### Actual behavior

The following exception is thrown:
```
$ java -jar target/quarkus-app/quarkus-run.jar 
__  ____  __  _____   ___  __ ____  ______ 
 --/ __ \/ / / / _ | / _ \/ //_/ / / / __/ 
 -/ /_/ / /_/ / __ |/ , _/ ,< / /_/ /\ \   
--\___\_\____/_/ |_/_/|_/_/|_|\____/___/   
2024-05-02 22:42:12,818 ERROR [io.qua.run.Application] (main) Failed to start application (with profile [prod]): java.lang.RuntimeException: Failed to start quarkus
	at io.quarkus.runner.ApplicationImpl.doStart(Unknown Source)
	at io.quarkus.runtime.Application.start(Application.java:101)
	at io.quarkus.runtime.ApplicationLifecycleManager.run(ApplicationLifecycleManager.java:111)
	at io.quarkus.runtime.Quarkus.run(Quarkus.java:71)
	at io.quarkus.runtime.Quarkus.run(Quarkus.java:44)
	at io.quarkus.runtime.Quarkus.run(Quarkus.java:124)
	at io.quarkus.runner.GeneratedMain.main(Unknown Source)
	at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:103)
	at java.base/java.lang.reflect.Method.invoke(Method.java:580)
	at io.quarkus.bootstrap.runner.QuarkusEntryPoint.doRun(QuarkusEntryPoint.java:62)
	at io.quarkus.bootstrap.runner.QuarkusEntryPoint.main(QuarkusEntryPoint.java:33)
Caused by: java.lang.NullPointerException: Cannot invoke "io.quarkus.runtime.RuntimeValue.getValue()" because "frameworkRouter" is null
	at io.quarkus.vertx.http.runtime.VertxHttpRecorder.finalizeRouter(VertxHttpRecorder.java:499)
	at io.quarkus.deployment.steps.VertxHttpProcessor$finalizeRouter2040953008.deploy_0(Unknown Source)
	at io.quarkus.deployment.steps.VertxHttpProcessor$finalizeRouter2040953008.deploy(Unknown Source)
	... 11 more
```

### How to Reproduce?

See example in this gist: https://gist.github.com/thomasdarimont/458dd8ce85dd347ddf3c690d61944559

### Output of `uname -a` or `ver`

Linux neumann 6.8.7-100.fc38.x86_64 #1 SMP PREEMPT_DYNAMIC Wed Apr 17 19:34:28 UTC 2024 x86_64 GNU/Linux

### Output of `java -version`

openjdk version "22.0.1" 2024-04-16 OpenJDK Runtime Environment Temurin-22.0.1+8 (build 22.0.1+8) OpenJDK 64-Bit Server VM Temurin-22.0.1+8 (build 22.0.1+8, mixed mode, sharing)

### Quarkus version or git rev

3.10.0

### Build tool (ie. output of `mvnw --version` or `gradlew --version`)

Apache Maven 3.9.6

### Additional information

This is also reproducible with Quarkus 3.8.3 currently used by Keycloak (24.0.3).

The problem is triggered here:
https://github.com/quarkusio/quarkus/blob/3.8.3/extensions/vertx-http/runtime/src/main/java/io/quarkus/vertx/http/runtime/VertxHttpRecorder.java#L471

but I think the cause for this is here:
https://github.com/quarkusio/quarkus/blob/3.8.3/extensions/vertx-http/deployment/src/main/java/io/quarkus/vertx/http/deployment/VertxHttpProcessor.java#L272.

The problem is that the `frameworkRouter` is not initialized, since the check `&& route.isRouterFramework()` returns `false` for the `non-application-root-path` (the router type errornously set to `APPLICATION` instead of `FRAMEWORK`).

If I set the proper Router type in `io.quarkus.vertx.http.deployment.RouteBuildItem#isRouterFramework` then the application can start fine.

--- a/extensions/vertx-http/deployment/src/test/java/io/quarkus/vertx/http/accesslog/AccessLogWithManagementInterfaceTestCase.java
+++ b/extensions/vertx-http/deployment/src/test/java/io/quarkus/vertx/http/accesslog/AccessLogWithManagementInterfaceTestCase.java
@@ -1,0 +1,92 @@
+package io.quarkus.vertx.http.accesslog;
+
+import java.util.function.Consumer;
+
+import jakarta.enterprise.event.Observes;
+import jakarta.inject.Singleton;
+
+import org.hamcrest.Matchers;
+import org.jboss.shrinkwrap.api.asset.StringAsset;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.extension.RegisterExtension;
+
+import io.quarkus.builder.BuildChainBuilder;
+import io.quarkus.builder.BuildContext;
+import io.quarkus.builder.BuildStep;
+import io.quarkus.test.QuarkusExtensionTest;
+import io.quarkus.vertx.http.deployment.HttpRootPathBuildItem;
+import io.quarkus.vertx.http.deployment.RouteBuildItem;
+import io.restassured.RestAssured;
+import io.vertx.core.Handler;
+import io.vertx.ext.web.RoutingContext;
+
+/**
+ * Reproducer for the management-interface variant of
+ * https://github.com/quarkusio/quarkus/issues/40420.
+ * <p>
+ * With the management interface enabled, the non-application (framework) routes move to the
+ * management interface, so no framework router exists on the main server. Enabling the access log
+ * together with a custom HTTP root path used to crash startup with a {@link NullPointerException}
+ * while attaching the access-log handler to the (null) framework router.
+ */
+public class AccessLogWithManagementInterfaceTestCase {
+
+    private static final String APP_PROPS = """
+            quarkus.management.enabled=true
+            quarkus.http.root-path=/app
+            quarkus.http.non-application-root-path=/q
+            quarkus.http.access-log.enabled=true
+            """;
+
+    @RegisterExtension
+    static final QuarkusExtensionTest config = new QuarkusExtensionTest()
+            .withApplicationRoot((jar) -> jar
+                    .addAsResource(new StringAsset(APP_PROPS), "application.properties")
+                    .addClasses(MyObserver.class))
+            .addBuildChainCustomizer(buildCustomizer());
+
+    static Consumer<BuildChainBuilder> buildCustomizer() {
+        return new Consumer<BuildChainBuilder>() {
+            @Override
+            public void accept(BuildChainBuilder builder) {
+                builder.addBuildStep(new BuildStep() {
+                    @Override
+                    public void execute(BuildContext context) {
+                        // An application route (classified APPLICATION_ROUTE), so no framework router is created.
+                        HttpRootPathBuildItem buildItem = context.consume(HttpRootPathBuildItem.class);
+                        context.produce(buildItem.routeBuilder()
+                                .route("hello")
+                                .handler(new MyHandler())
+                                .blockingRoute()
+                                .build());
+                    }
+                }).produces(RouteBuildItem.class)
+                        .consumes(HttpRootPathBuildItem.class)
+                        .build();
+            }
+        };
+    }
+
+    public static class MyHandler implements Handler<RoutingContext> {
+        @Override
+        public void handle(RoutingContext routingContext) {
+            routingContext.response()
+                    .setStatusCode(200)
+                    .end(routingContext.request().path());
+        }
+    }
+
+    @Test
+    public void testApplicationStartsWithManagementInterfaceAndAccessLog() {
+        // The application boots (no NPE) and the application route is reachable under the HTTP root path.
+        RestAssured.given().get("/hello")
+                .then().statusCode(200).body(Matchers.equalTo("/app/hello"));
+    }
+
+    @Singleton
+    static class MyObserver {
+        void test(@Observes String event) {
+            // Do nothing
+        }
+    }
+}

--- a/extensions/vertx-http/deployment/src/test/java/io/quarkus/vertx/http/accesslog/AccessLogWithNonApplicationRootPathTestCase.java
+++ b/extensions/vertx-http/deployment/src/test/java/io/quarkus/vertx/http/accesslog/AccessLogWithNonApplicationRootPathTestCase.java
@@ -1,0 +1,146 @@
+package io.quarkus.vertx.http.accesslog;
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+import java.io.ByteArrayOutputStream;
+import java.io.IOException;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.util.Properties;
+import java.util.concurrent.TimeUnit;
+import java.util.function.Consumer;
+import java.util.function.Supplier;
+
+import jakarta.enterprise.event.Observes;
+import jakarta.inject.Singleton;
+
+import org.awaitility.Awaitility;
+import org.eclipse.microprofile.config.inject.ConfigProperty;
+import org.hamcrest.Matchers;
+import org.jboss.shrinkwrap.api.ShrinkWrap;
+import org.jboss.shrinkwrap.api.asset.ByteArrayAsset;
+import org.jboss.shrinkwrap.api.spec.JavaArchive;
+import org.junit.jupiter.api.AfterEach;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.extension.RegisterExtension;
+
+import io.quarkus.bootstrap.util.IoUtils;
+import io.quarkus.builder.BuildChainBuilder;
+import io.quarkus.builder.BuildContext;
+import io.quarkus.builder.BuildStep;
+import io.quarkus.test.QuarkusExtensionTest;
+import io.quarkus.vertx.http.deployment.HttpRootPathBuildItem;
+import io.quarkus.vertx.http.deployment.RouteBuildItem;
+import io.restassured.RestAssured;
+import io.vertx.core.Handler;
+import io.vertx.ext.web.RoutingContext;
+
+/**
+ * Reproducer for https://github.com/quarkusio/quarkus/issues/40420.
+ * <p>
+ * When the access log is enabled together with a custom non-application root path that is disjoint
+ * from the HTTP root path, and no framework router was ever created (no route is classified as a
+ * framework route), startup used to fail with a {@link NullPointerException} while attaching the
+ * access-log handler to the (null) framework router.
+ * <p>
+ * The application must start and the access log of application routes must keep working.
+ */
+public class AccessLogWithNonApplicationRootPathTestCase {
+
+    @RegisterExtension
+    static final QuarkusExtensionTest config = new QuarkusExtensionTest()
+            .setArchiveProducer(new Supplier<>() {
+                @Override
+                public JavaArchive get() {
+                    try {
+                        Path logDirectory = Files.createTempDirectory("quarkus-tests");
+                        Properties p = new Properties();
+                        p.setProperty("quarkus.http.root-path", "/auth");
+                        p.setProperty("quarkus.http.non-application-root-path", "/q");
+                        p.setProperty("quarkus.http.access-log.enabled", "true");
+                        p.setProperty("quarkus.http.access-log.log-to-file", "true");
+                        p.setProperty("quarkus.http.access-log.base-file-name", "server");
+                        p.setProperty("quarkus.http.access-log.log-directory", logDirectory.toAbsolutePath().toString());
+                        ByteArrayOutputStream out = new ByteArrayOutputStream();
+                        p.store(out, null);
+                        return ShrinkWrap.create(JavaArchive.class)
+                                .add(new ByteArrayAsset(out.toByteArray()), "application.properties")
+                                .addClasses(MyObserver.class);
+                    } catch (IOException e) {
+                        throw new RuntimeException(e);
+                    }
+                }
+            })
+            .addBuildChainCustomizer(buildCustomizer());
+
+    static Consumer<BuildChainBuilder> buildCustomizer() {
+        return new Consumer<BuildChainBuilder>() {
+            @Override
+            public void accept(BuildChainBuilder builder) {
+                builder.addBuildStep(new BuildStep() {
+                    @Override
+                    public void execute(BuildContext context) {
+                        // An application route (classified APPLICATION_ROUTE), so no framework router is created.
+                        HttpRootPathBuildItem buildItem = context.consume(HttpRootPathBuildItem.class);
+                        context.produce(buildItem.routeBuilder()
+                                .route("hello")
+                                .handler(new MyHandler())
+                                .blockingRoute()
+                                .build());
+                    }
+                }).produces(RouteBuildItem.class)
+                        .consumes(HttpRootPathBuildItem.class)
+                        .build();
+            }
+        };
+    }
+
+    // Injected at runtime so the value survives the QuarkusUnitTest classloader boundary
+    // (a static field set by the archive producer would not be visible to the test method).
+    @ConfigProperty(name = "quarkus.http.access-log.log-directory")
+    Path logDirectory;
+
+    @BeforeEach
+    public void before() throws IOException {
+        Files.createDirectories(logDirectory);
+    }
+
+    @AfterEach
+    public void after() throws IOException {
+        IoUtils.recursiveDelete(logDirectory);
+    }
+
+    public static class MyHandler implements Handler<RoutingContext> {
+        @Override
+        public void handle(RoutingContext routingContext) {
+            routingContext.response()
+                    .setStatusCode(200)
+                    .end(routingContext.request().path());
+        }
+    }
+
+    @Test
+    public void testAccessLogWithDisjointNonApplicationRootPath() {
+        // The application boots (no NPE) and the application route is reachable under the HTTP root path.
+        // RestAssured is configured with the configured HTTP root path (/auth) as its base path.
+        RestAssured.given().get("/hello")
+                .then().statusCode(200).body(Matchers.equalTo("/auth/hello"));
+
+        // The request to the application route is written to the access log.
+        Awaitility.given().pollInterval(100, TimeUnit.MILLISECONDS)
+                .atMost(10, TimeUnit.SECONDS)
+                .untilAsserted(() -> {
+                    Path logFile = logDirectory.resolve("server.log");
+                    assertThat(Files.exists(logFile)).isTrue();
+                    assertThat(Files.readString(logFile)).contains("/auth/hello");
+                });
+    }
+
+    @Singleton
+    static class MyObserver {
+        void test(@Observes String event) {
+            // Do nothing
+        }
+    }
+}

--- a/extensions/vertx-http/runtime/src/main/java/io/quarkus/vertx/http/runtime/VertxHttpRecorder.java
+++ b/extensions/vertx-http/runtime/src/main/java/io/quarkus/vertx/http/runtime/VertxHttpRecorder.java
@@ -653,10 +653,21 @@ public class VertxHttpRecorder {
         } else if (nonRootPath.startsWith(rootPath)) {
             httpRouteRouter.route().order(RouteConstants.ROUTE_ORDER_ACCESS_LOG_HANDLER).handler(handler);
         } else if (rootPath.startsWith(nonRootPath)) {
-            frameworkRouter.getValue().route().order(RouteConstants.ROUTE_ORDER_ACCESS_LOG_HANDLER).handler(handler);
+            // The framework (non-application) routes live on a dedicated router mounted above the
+            // application router. If that router was never created (e.g. all non-application routes
+            // moved to the management interface, or none were classified as framework routes), fall
+            // back to the application router so application routes are still logged.
+            if (frameworkRouter != null) {
+                frameworkRouter.getValue().route().order(RouteConstants.ROUTE_ORDER_ACCESS_LOG_HANDLER).handler(handler);
+            } else {
+                httpRouteRouter.route().order(RouteConstants.ROUTE_ORDER_ACCESS_LOG_HANDLER).handler(handler);
+            }
         } else {
             httpRouteRouter.route().order(RouteConstants.ROUTE_ORDER_ACCESS_LOG_HANDLER).handler(handler);
-            frameworkRouter.getValue().route().order(RouteConstants.ROUTE_ORDER_ACCESS_LOG_HANDLER).handler(handler);
+            // Only attach to the framework router when it exists; see comment above.
+            if (frameworkRouter != null) {
+                frameworkRouter.getValue().route().order(RouteConstants.ROUTE_ORDER_ACCESS_LOG_HANDLER).handler(handler);
+            }
         }
     }
 

